doc/dev/crimson/pipelines.rst

```diff
@@ -0,0 +1,206 @@
+                                                       |cBLU| recovery +-+             +----+---------+
+                                                       +----+----------+ |             |cBLU| process |
+                                                                         |             +----+---------+
+                                                                         V
+                                                                       +----+---------+
+                                                                       |cBLU| obj ctx +-+
+                                                                       +----+---------+ |
+                                                                                        |
+                                                                                        V
+                                                                                       +----+---------+
+                                                                                       |cBLU| process |
+                                                                                       +----+---------+
+
+In the diagram above, two requests from the same client are being served in parallel. They share the same
+pipeline. But before one transits to the next stage, it have to wait until another leaves that stage.
+
+Better Throughput
+=================
+
+Some of us might argue, can we improve the throughput using some well known techniques?
+
+Reordering
+----------
+
+When serving requests from a client performing asynchronous I/O, OSD has the opportunity to execute
+multiple requests from the same client in parallel. But due to the second restriction, we should always
+reply the client in-order. But this does not prevent us from **executing** the requests ouf-of-order, as
+long as the reordering does not change the behavior of the execution or breaks the dependencies between
+requests.
+
+- for read requests, the "behavior" is the returned metadata and payload of object.
+- for write requests, the "behavior" is the state transition of the object storage of this PG shard and its peers.
+
+Let's consider a use case of read-read reordering, where the client sends two read requests targeting the
+same object in a row. Transposing these two read operation does not change the corresponding replies, unless
+there is a write request sneaked in after the first read request is served. But since the client sending
+read requests is not the one sending write request, the former should not be annoyed if the reordering
+changes the responses -- the object state is still consistent, and the contract of RADOS protocol is not
+broken.
+
+When it comes to read-write reordering, we cannot ensure the consistency of the behavior without checking
+if the extents of read and write operations overlap with each other. The same applies to write-write reordering.
+
```

Comment:
It's worth noting that the single largest reason not to worry about any of this is that (cephfs, rbd, radosgw) don't work in terms of a single huge rados object, they use large collections of bounded size rados objects.  10 request on an rbd device won't be pipelined in general because *we already split them among different rados objects based on offset*.